File: 0120-triangle/0120-triangle.js
/**
 * @param {number[][]} triangle
 * @return {number}
 */
var minimumTotal = function(triangle) {
    // Space Optimized Bottom Up Approach;
    // If we are goingg from 0 to n in Top Down, then
    // we need to go from n to 0 in Bottom up!

    let rowsLen = triangle.length;
    let prevDP = new Array(rowsLen).fill(0);
    
    // console.log(prevDP);
    // Base Case:
    // prevDP[0] = triangle[0][0];
    for(let colIdx = 0; colIdx < rowsLen; colIdx++){
        prevDP[colIdx] = triangle[rowsLen - 1][colIdx];
    }

    for(let rowIdx = rowsLen - 2; rowIdx >= 0; rowIdx--){
        const currDP = new Array(rowIdx + 1).fill(0);
        // console.log(currDP.length, rowIdx, currDP);

        for(colIdx = rowIdx; colIdx >= 0; colIdx--){
            // Choose the currentIdx 
            const choseCurrIdx = triangle[rowIdx][colIdx] + prevDP[colIdx];

            // Choose the nextIdx
            const choseNextIdx = triangle[rowIdx][colIdx] + prevDP[colIdx + 1];

            currDP[colIdx] = Math.min(choseCurrIdx, choseNextIdx);
        }

        prevDP = currDP;
    }   

    // console.log(memoDP);
    return prevDP[0];
};

File: 0120-triangle/0120-triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        # Top Down Approach:
        m_rows = len(triangle)

        memo_dp = [[-1 for _ in range(m_rows)] for _ in range(m_rows)]

        def all_path_sum(curr_idx, curr_row_len):
            # print(curr_idx, curr_row_len)
            # Base Case:
            if curr_idx == 0 and curr_row_len == 0:
                return triangle[curr_idx][curr_row_len]
            elif curr_idx < 0 or curr_row_len < 0 or curr_idx > curr_row_len:
                return float('inf')
            
            if memo_dp[curr_idx][curr_row_len] != -1:
                return memo_dp[curr_idx][curr_row_len]
            
            # pick the current_idx;
            chose_curr_idx = triangle[curr_row_len][curr_idx] + all_path_sum(curr_idx, curr_row_len - 1)

            chose_next_idx = triangle[curr_row_len][curr_idx] + all_path_sum(curr_idx - 1, curr_row_len - 1)

            memo_dp[curr_idx][curr_row_len] = min(chose_curr_idx, chose_next_idx)
            return memo_dp[curr_idx][curr_row_len]
        
        min_path_sum = float('inf')
        for idx_i in range(m_rows):
           new_min_path_sum = all_path_sum(idx_i, m_rows - 1)
           min_path_sum = min(min_path_sum, new_min_path_sum)

        return min_path_sum
        